File: algorithms/1_easy/R/01576_replace_all_s_to_avoid_consecutive_repeating_characters/solution.cpp
class Solution {
public:
    string modifyString(string s) {
        const size_t len{s.size()};

        for (size_t i{0}; i < len; ++i) {
            if (s[i] == '?') {
                bool a_allowed{true};
                bool b_allowed{true};

                if ((i > 0 && s[i - 1] == 'a') || (i + 1 < len && s[i + 1] == 'a')) {
                    a_allowed = false;
                }

                if ((i > 0 && s[i - 1] == 'b') || (i + 1 < len && s[i + 1] == 'b')) {
                    b_allowed = false;
                }

                if (a_allowed) {
                    s[i] = 'a';
                } else if (b_allowed) {
                    s[i] = 'b';
                } else {
                    s[i] = 'c';
                }
            }
        }

        return s;
    }
};

File: algorithms/1_easy/R/01576_replace_all_s_to_avoid_consecutive_repeating_characters/solution.rs
impl Solution {
    pub fn modify_string(s: String) -> String {
        let mut bytes = s.into_bytes();
        let len = bytes.len();

        for i in 0..len {
            if bytes[i] == b'?' {
                let mut a_allowed = true;
                let mut b_allowed = true;

                if (i > 0 && bytes[i - 1] == b'a') || (i + 1 < len && bytes[i + 1] == b'a') {
                    a_allowed = false;
                }

                if (i > 0 && bytes[i - 1] == b'b') || (i + 1 < len && bytes[i + 1] == b'b') {
                    b_allowed = false;
                }

                if a_allowed {
                    bytes[i] = b'a';
                } else if b_allowed {
                    bytes[i] = b'b';
                } else {
                    bytes[i] = b'c';
                }
            }
        }

        String::from_utf8(bytes).unwrap()
    }
}
